src/matrices/skeletonization.jl

```diff
@@ -0,0 +1,394 @@
+
+    # go through all lines stored in the sorted line matrix
+    @inbounds for l_ind = 1:Nlines1
+        # check if line is still in the rendered line matrix
+        common_ind = intersect(findall(x-> x==true, lines1t[l_ind] .== lines_copy1t),
+                        findall(x-> x==true, lines1l[l_ind] .== lines_copy1l),
+                        findall(x-> x==true, lines1c[l_ind] .== lines_copy1c))
+        isempty(common_ind) ? continue : nothing
+        # get index pair for start of the line
+        linei, columni = lines1l[l_ind], lines1c[l_ind]
+        # copy this line in the final line matrix
+        push!(lines_final_t, lines1t[l_ind])
+        push!(lines_final_l, lines1l[l_ind])
+        push!(lines_final_c, lines1c[l_ind])
+        # delete this line from the RP
+        delete_line_from_cl_ret_RP!(X_hori1, lines1t[l_ind], lines1l[l_ind], lines1c[l_ind])
+        # go along each point of the line and check for neighbours
+        l_max = lines1t[l_ind]
+
+        @inbounds for l = 1:l_max
+            # scan each line twice - above and underneth
+            for index = -1:2:1
+                # make sure not to exceed RP-boundaries
+                (linei+index > N || linei+index == 0) ? break : nothing
+                # if there is a neighbouring point, call recursive scan-function
+                (X_hori1[linei+index, columni+l-1] != 0) ? scan_lines!(X_hori1, lines_copy1t,
+                                lines_copy1l, lines_copy1c, linei+index, columni+l-1) : nothing
+            end
+        end
+    end
+    return lines_final_t, lines_final_l, lines_final_c
+end
+
+# compute a `3-by-N` matrix, which stores all horizontal lines from a converted
+# recurrence matrix (after applying `convert_recurrence_matrix` to a ARM). The
+# line lengths of each found line are stored in the first row, and the starting
+# indices of its row and column, are stored in the 2nd and 3rd row, respectively.
+function horizontalhisto(R::SparseMatrixCSC)
+
+    N = size(R)[1]
+    liness = zeros(Int, 3, 1)
+    @inbounds for j = 1:N
+        starts = findall(diff([0; @view R[j,:]]).==1)
+        ends = findall(diff([@view R[j,:]; 0]).==-1)
+
+        if ~isempty(starts)
+            lines = zeros(Int, 3, length(starts))
+            for n=1:length(starts)
+                lines[2,n] = j
+                lines[3,n] = starts[n]
+                lines[1,n] = ends[n] - starts[n] + 1
+            end
+            liness = hcat(liness,lines)
+        end
+    end
+    # remove lines of length zero (=no line)
+    no_zero_lines = findall(liness[1,:].!=0)
+    liness = liness[:, no_zero_lines]
+    ls = size(liness)
+    liness = liness[:,sortperm(@view liness[1, :]; rev=true)]
+    return vec(liness[1,:]), vec(liness[2,:]), vec(liness[3,:])
+end
+
+# function horizontalhisto2(R::SparseMatrixCSC)
+#     # Transpose R in order to get the "horizontal" lines
+#     # rows = colvals(R[2:end,:])
+#     # cols = rowvals(R[2:end,:])
```

Comment:
What's the purpose of this huge block of commented code?

Reply:
This function serves the same purpose as `horizontalhisto()` does. In contrast to the latter, in this approach I attempted to make it faster than `horizontalhisto()`. This block is so huge, because it actually copies `_linehistograms` a lot, which is the very performant core-function for RQA by @heliosdrm . After our little discussion in #98 I thought it might be useful to reuse that code and compute the rotated lines in the skeletonization method. It turned out that it is in fact slower for larger input (see my posted figure above). I left it for @heliosdrm , maybe he might be able to make it work faster. For the actual implementation it is not necessary.

Regarding your clean-code-isssues, @Datseris and @heliosdrm : I am very thankful for any suggestions to make it cleaner and look nicer. Just let me know or simply do it, if it is obvious. I do not care personally if we merge it now or later after cleaning.

Reply:
Ok, I'll give a closer look to the two versions of horizontal histograms and see if I can find some improvement. Would you be so kind to provide a script to reproduce the benchmarking above?

I'll also propose "cleaner" versions of the code.

Reply:
Sure! Try something like this AFTER outcommenting `horizontalhisto2()`:
```
using RecurrenceAnalysis
using LinearAlgebra
using TimerOutputs

using PyPlot
pygui(true)

data = sin.(2*π .* (0:5500)./ 60)
Y = embed(data, 3, 15)

timings_my_way = zeros(19)
timings_based_on_linehistograms = zeros(19)
for (i,N) = enumerate(500:250:5000)
        println("i is: $i")
        println(N)
        X = RecurrenceMatrix(Y[1:N,:], 0.25; fixedrate=true)
        X_cl1 = RecurrenceAnalysis.convert_recurrence_matrix(tril(X))
        timings_my_way[i] = @elapsed RecurrenceAnalysis.horizontalhisto(X_cl1)
        timings_based_on_linehistograms[i] = @elapsed RecurrenceAnalysis.horizontalhisto2(X_cl1)
end


fig = figure(figsize = (10,5))

plot(vec(500:250:5000), timings_my_way, color="red", label="actual method")
plot(vec(500:250:5000), timings_based_on_linehistograms, color="blue", label="method based on _linehistograms")
ylabel("computation time")
xlabel("trajectory length")
grid()
legend()
```

Thanks @heliosdrm for taking a closer look. Also looking forward to your suggested ways of eliminating the ugly `ifelse`-blocks, @Datseris and @heliosdrm .